interceptor/intercept.go

```diff
@@ -85,66 +89,179 @@ func (i *Interceptor) Intercept(nextHop string, reqPaymentHash []byte, reqOutgoi
+				return InterceptResult{
+					Action:      INTERCEPT_FAIL_HTLC_WITH_CODE,
+					FailureCode: FAILURE_TEMPORARY_CHANNEL_FAILURE,
+				}, nil
+			}
+
+			if !isConnected {
+				// If not connected, send a notification to the registered
+				// notification service for this client if available.
+				notified, err := i.notificationService.Notify(
+					hex.EncodeToString(nextHop),
+					reqPaymentHashStr,
+				)
+
+				// If this errors or the client is not notified, the client
+				// is offline or unknown. We'll resume the HTLC (which will
+				// result in UNKOWN_NEXT_PEER)
 				if err != nil {
-					log.Printf("time.Parse(%s, %s) failed. Failing channel open: %v", basetypes.TIME_FORMAT, params.ValidUntil, err)
 					return InterceptResult{
-						Action:      INTERCEPT_FAIL_HTLC_WITH_CODE,
-						FailureCode: FAILURE_TEMPORARY_CHANNEL_FAILURE,
+						Action: INTERCEPT_RESUME,
 					}, nil
 				}
 
-				if time.Now().UTC().After(validUntil) {
-					if !i.isCurrentChainFeeCheaper(token, params) {
-						log.Printf("Intercepted expired payment registration. Failing payment. payment hash: %x, valid until: %s", paymentHash, params.ValidUntil)
+				if notified {
+					log.Printf("Notified %x of pending htlc", nextHop)
+					d, err := time.ParseDuration(i.config.NotificationTimeout)
+					if err != nil {
+						log.Printf("WARN: No NotificationTimeout set. Using default 1m")
+						d = time.Minute
+					}
+					timeout := time.Now().Add(d)
+					err = i.client.WaitOnline(nextHop, timeout)
+
+					// If there's an error waiting, resume the htlc. It will
+					// probably fail with UNKNOWN_NEXT_PEER.
+					if err != nil {
+						log.Printf(
+							"waiting for peer %x to come online failed with %v",
+							nextHop,
+							err,
+						)
 						return InterceptResult{
-							Action:      INTERCEPT_FAIL_HTLC_WITH_CODE,
-							FailureCode: FAILURE_TEMPORARY_CHANNEL_FAILURE,
+							Action: INTERCEPT_RESUME,
 						}, nil
 					}
 
-					log.Printf("Intercepted expired payment registration. Opening channel anyway, because it's cheaper at the current rate. paymenthash: %s, params: %+v", reqPaymentHashStr, params)
+					log.Printf("Peer %x is back online. Continue htlc.", nextHop)
+					// At this point we know a few things.
+					// - This is either a channel partner or a registered payment
+					// - they were offline
+					// - They got notified about the htlc
+					// - They came back online
+					// So if this payment was not registered, this is a channel
+					// partner and we have to wait for the channel to become active
+					// before we can forward.
+					if !isRegistered {
+						err = i.client.WaitChannelActive(nextHop, timeout)
+						if err != nil {
+							log.Printf(
+								"waiting for channnel with %x to become active failed with %v",
+								nextHop,
+								err,
+							)
+							return InterceptResult{
+								Action: INTERCEPT_RESUME,
+							}, nil
+						}
+					}
+				} else if isProbe {
```

Comment:
The following two cases (else branches)  seems unrelated to the notification itself. Perhaps it is better to handle them after the notification code block (that start at` if (!isConnected) `)?
It seems that it will also allow extracting out the notification specific logic to its own function making the intercept logic easier to follow.

Reply:
The probe case is the hardest part of this flow. Maybe I need to know exactly when we want to return INCORRECT_OR_UNKNOWN_PAYMENT_DETAILS in case the payment hash is a probe payment hash.
Does this make sense?
- Payment is registered:
  - Peer is offline: Yes 
  - Peer is online:
    - Channel has not been opened: Yes
    - Channel has been opened: No

And when do we notify the client to come online when there is a probe payment hash? Never?